### PR TITLE
Features/pastushkov code editor

### PR DESCRIPTION
**Pull request by opastushkov**
Hotkey Ctrl + / (singleline and multiline comment/uncomment).

### CODE REVIEW

File: src/editor/lexercpp.cpp [hunk 1]
```diff
@@ -1,6 +1,212 @@
 #include "lexercpp.h"
+#include <QDebug>
 
 LexerCPP::LexerCPP()
```
Comment (PeterLysyk) [inline src/editor/lexercpp.cpp]: I suggest better idea would be to use the word "default"

File: src/editor/lexercpp.cpp [hunk 2]
```diff
@@ -1,6 +1,212 @@
...
 {
 
 }
+
+LexerCPP::~LexerCPP()
```
Comment (PeterLysyk) [inline src/editor/lexercpp.cpp]: I suggest better idea would be to use the word "default"

File: src/editor/lexercpp.cpp [hunk 3]
```diff
@@ -1,6 +1,212 @@
...
+                else if(isDigit(sym))
+                    changeState(NUM, sym);
+                else if(isOperator(QString(sym)))
+                    changeState(OPER, sym);
+                else if(sym == '\'' || sym == '"')
+                    changeState(LIT, sym);
+            break;
+
+            case ID:
+                if(isSpace(sym) || isOperator((QString)sym) || sym == '\"' || sym == '\'')
```
Comment (PeterLysyk) [inline src/editor/lexercpp.cpp]: use (  )  for each condition like
if((isSpace(sym)) || (isOperator((QString)sym)) || (sym == '\"') || (sym == '\''))

File: src/editor/lexercpp.cpp [hunk 4]
```diff
@@ -1,6 +1,212 @@
...
+                else if(isSpace(sym) || isOperator((QString)sym) || sym == '\"' || sym == '\'')
+                    addLexem();
+                else if(isNumber(current_lexem + sym))
+                    current_lexem += sym;
+                else
+                    state = UNDEF;
+            break;
+
+            case FNUM:
+                if(isSpace(sym) || isOperator((QString)sym) || sym == '\"' || sym == '\'')
```
Comment (PeterLysyk) [inline src/editor/lexercpp.cpp]: maybe better idea is to create function for this check. You do the same as in the 149 and 162 lines

File: src/editor/spaces.h [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#ifndef SPACES_H
+#define SPACES_H
+#include <QSet>
+
+const QSet<QChar> SPACES = {'\n', ' ', '\r', '\t'};
```
Comment (PeterLysyk) [inline src/editor/spaces.h:5]: what about "\v"?

**Review by PeterLysyk — APPROVED** [review]
approved

File: src/editor/fmstates.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#ifndef FMSTATES_H
+#define FMSTATES_H
+
+enum States {
```
Comment (DANIV18) [inline src/editor/fmstates.h]: Use enum class

File: src/editor/highlightercpp.cpp [hunk 7]
```diff
@@ -1,6 +1,33 @@
...
+#include <QDebug>
 
-/*Highlightercpp::Highlightercpp(CodeEditor *parent):QSyntaxHighlighter (parent)
+Highlightercpp::Highlightercpp(QTextDocument* parent):QSyntaxHighlighter (parent)
 {
 
-}*/
+}
+
+void Highlightercpp::setData(QVector<Token> _data)
```
Comment (DANIV18) [inline src/editor/highlightercpp.cpp:9]: Why you don't use const QVector<Token> &data?

File: src/editor/highlightercpp.h [hunk 8]
```diff
@@ -1,11 +1,19 @@
 #ifndef HIGHLIGHTERCPP_H
 #define HIGHLIGHTERCPP_H
 #include <QSyntaxHighlighter>
+#include <QTextCharFormat>
+#include <token.h>
 
 class Highlightercpp: QSyntaxHighlighter
 {
+private:
+    unsigned int currentLine;
```
Comment (DANIV18) [inline src/editor/highlightercpp.h:10]: class members names must start with m prefix

File: src/editor/highlightercpp.h [hunk 9]
```diff
@@ -1,11 +1,19 @@
...
+    QVector<Token> data;
 public:
-   // Highlightercpp();
+    Highlightercpp(QTextDocument* parent);
+    void setData(QVector<Token>);
```
Comment (DANIV18) [inline src/editor/highlightercpp.h:14]: Why you don't use const QVector &data?

File: src/editor/lexercpp.cpp [hunk 10]
```diff
@@ -1,6 +1,212 @@
...
 {
 
 }
+
+LexerCPP::~LexerCPP()
+{
+
+}
+
+bool LexerCPP::isKeyword(QString lexem)
```
Comment (DANIV18) [inline src/editor/lexercpp.cpp]: use const QString &lexem

File: src/editor/operators.h [hunk 11]
```diff
@@ -0,0 +1,14 @@
+#ifndef OPERATORS_H
+#define OPERATORS_H
+#include <QSet>
+
+const QSet<QString> OPERATORS = {"=", "+", "-", "*", "/", "%",
```
Comment (DANIV18) [inline src/editor/operators.h]: all global sets you can move to specific namespace

File: src/editor/type.h [hunk 12]
```diff
@@ -0,0 +1,15 @@
+#ifndef TYPE_H
+#define TYPE_H
+
+enum Type {
```
Comment (DANIV18) [inline src/editor/type.h]: use enum class

**Review by anantonyk — APPROVED** [review]

**Pull request merged**